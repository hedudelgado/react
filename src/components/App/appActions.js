import { HeaderInterface, InvalidSessionErrorCodeMap, InvalidSessionHelper } from 'mega-code';
import {
  getCountries,
  getNewsletterRegions,
  getNewsletterPreferences,
  getProfile,
  updateProfile,
  deleteCustomer,
} from '../../utils/api';

// actions
export const USER_LOGGED_IN_SUCCESSFULLY = 'USER_LOGGED_IN_SUCCESSFULLY';
export const COUNTRIES_RECEIVED = 'COUNTRIES_RECEIVED';
export const PROFILE_RECEIVED = 'PROFILE_RECEIVED';
export const REGIONS_RECEIVED = 'REGIONS_RECEIVED';
export const NEWSLETTER_RECEIVED = 'NEWSLETTER_RECEIVED';
export const SET_EDIT_VIEW = 'SET_EDIT_VIEW';
export const SET_NOTIFICATION_MESSAGE = 'SET_NOTIFICATION_MESSAGE';
export const PROFILE_UPDATED = 'PROFILE_UPDATED';
export const NEWSLETTER_UPDATED = 'NEWSLETTER_UPDATED';
export const RESET_NEWSLETTER_INFO = 'RESET_NEWSLETTER_INFO';
export const RESET_USER_INFO = 'RESET_USER_INFO';
// views
export const DEFAULT_VIEW = undefined;
export const USER_DETAILS_VIEW = 'USER_DETAILS_VIEW';
export const PASSWORD_VIEW = 'PASSWORD_VIEW';
export const PAYMENT_CARD_VIEW = 'PAYMENT_CARD_VIEW';
export const ADD_GUEST_VIEW = 'ADD_GUEST_VIEW';
export const REGULAR_GUESTS_VIEW = 'REGULAR_GUESTS_VIEW';
export const NEWSLETTER_VIEW = 'NEWSLETTER_VIEW';
export const ROOM_REQUIREMENTS_VIEW = 'ROOM_REQUIREMENTS_VIEW';
export const EXTRAS_PREFERENCES_VIEW = 'EXTRAS_PREFERENCES_VIEW';
export const MEMORABLE_WORD_VIEW = 'MEMORABLE_WORD_VIEW';
// errors
export const PROFILE_ERROR = 'PROFILE_ERROR';
export const PASSWORD_ERROR = 'PASSWORD_ERROR';
export const NEWSLETTER_ERROR = 'NEWSLETTER_ERROR';
export const PAYMENT_CARD_ERROR = 'PAYMENT_CARD_ERROR';
export const USER_DETAILS_ERROR = 'USER_DETAILS_ERROR';
export const REGULAR_GUESTS_ERROR = 'REGULAR_GUESTS_ERROR';
export const MEMORABLE_WORD_ERROR = 'MEMORABLE_WORD_ERROR';
export const ROOM_REQUIREMENTS_ERROR = 'ROOM_REQUIREMENTS_ERROR';
export const EXTRAS_PREFERENCES_ERROR = 'EXTRAS_PREFERENCES_ERROR';

const checkForInvalidSession = (error, config, dispatch) => {
  if (InvalidSessionHelper
    .isSessionExpired(error, config)) {
    dispatch({
      type: RESET_USER_INFO,
    });
    dispatch({
      type: RESET_NEWSLETTER_INFO,
    });
    HeaderInterface.userLogOut(HeaderInterface.LOG_OUT_REASONS.SESSION_EXPIRED);
  }
};

export const getSettings = (email, sessionId, business) => (dispatch, getState) => {
  getProfile(email, sessionId, business).then((profile) => {
    dispatch({
      type: PROFILE_RECEIVED,
      profile,
    });
  }).catch((error) => {
    checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
    dispatch({
      type: PROFILE_ERROR,
      error,
    });
  });
  getCountries()
    .then((countries) => {
      dispatch({
        type: COUNTRIES_RECEIVED,
        countries,
      });
    });
  if (getState().application.showEmailPreferences) {
    getNewsletterRegions()
      .then((regions) => {
        dispatch({
          type: REGIONS_RECEIVED,
          regions,
        });
      });

    getNewsletterPreferences(email)
      .then((newsletter) => {
        dispatch({
          type: NEWSLETTER_RECEIVED,
          newsletter,
        });
      })
      .catch(e => dispatch({
        type: NEWSLETTER_ERROR,
        error: e,
      }));
  }
};

export const updateUserProfile = payload => dispatch => updateProfile(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
  return Promise.reject(error);
});

export const deleteUserProfile = payload => dispatch => deleteCustomer(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.DELETE_PROFILE.invalidSession, dispatch);
  return Promise.reject(error);
});

export default {
  getSettings,
};
