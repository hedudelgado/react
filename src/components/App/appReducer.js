import {
  SET_NOTIFICATION_MESSAGE, SET_EDIT_VIEW, DEFAULT_VIEW,
} from './appActions';

const defaultState = {
  editViewName: DEFAULT_VIEW,
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case SET_EDIT_VIEW:
      return {
        ...state,
        editViewName: action.editViewName,
      };
    case SET_NOTIFICATION_MESSAGE:
      return {
        ...state,
        notificationMessage: action.notificationMessage,
      };
    default:
      return state;
  }
};
