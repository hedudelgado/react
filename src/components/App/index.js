export { default } from './App';

export { default as AppActions } from './appActions';

export { default as AppContainer } from './AppContainer';

export { default as modalReducer } from './modalReducer';
