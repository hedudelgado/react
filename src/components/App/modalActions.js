export const MODAL_OPEN = 'MODAL_OPEN';
export const MODAL_CLOSE = 'MODAL_CLOSE';

export const openModal = content => ({
  type: MODAL_OPEN,
  content,
});

export const closeModal = () => ({
  type: MODAL_CLOSE,
});
