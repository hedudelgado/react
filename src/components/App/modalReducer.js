import {
  MODAL_OPEN,
  MODAL_CLOSE,
} from './modalActions';

export default (state = {
  display: false,
  content: null,
}, action) => {
  const {
    type,
    content,
  } = action;

  switch (type) {
    case MODAL_OPEN:
      return {
        ...state,
        display: true,
        content,
      };
    case MODAL_CLOSE:
      return {
        ...state,
        display: false,
        content: null,
      };
    default:
      return state;
  }
};
