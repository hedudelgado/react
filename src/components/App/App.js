import React, { Component } from 'react';
import {
  array,
  object,
  func,
  bool,
  node,
} from 'prop-types';
import {
  PASSWORD_VIEW,
  PAYMENT_CARD_VIEW,
  USER_DETAILS_VIEW,
  ROOM_REQUIREMENTS_VIEW,
  EXTRAS_PREFERENCES_VIEW,
} from './appActions';
import { ComponentTogglerContainer } from '../ComponentToggler';
import UserProfile from '../UserProfile';
import Password from '../Password';
import UserProfileFormContainer from '../UserProfileForm';
import PasswordFormContainer from '../PasswordForm';
import { PaymentCardContainer } from '../PaymentCard';
import { RegularGuestsContainer } from '../RegularGuests';
import { NewsletterContainer } from '../Newsletter';
import Modal from '../../layout/Modal';
import PaymentCardFormContainer from '../PaymentCardForm';
import { DeleteCustomerContainer } from '../DeleteCustomer';
import RoomRequirements from '../RoomRequirements';
import RoomRequirementsFormContainer from '../RoomRequirementsForm';
import ExtrasPreferences from '../ExtrasPreferences';
import ExtrasPreferencesFormContainer from '../ExtrasPreferencesForm';
import BaseStyle from '../../layout/BaseStyle';
import MemorableWord from '../MemorableWord';
import MoveAccount from '../MoveAccount';
import {
  STAYER, SELF, managePermissionFlag, USER_LOGGED_IN_EVENT,
} from '../../config.json';
import ManagePermission from '../ManagePermission';

class App extends Component {
  constructor(props) {
    super(props);
    this.handleLoggedIn = this.handleLoggedIn.bind(this);
  }

  componentDidMount() {
    window.addEventListener('message', this.handleLoggedIn);
  }

  componentWillUnmount() {
    window.removeEventListener('message', this.handleLoggedIn);
  }

  async handleLoggedIn(event) {
    if (event.data.action === USER_LOGGED_IN_EVENT) {
      const { getSettings } = this.props;
      const { contactDetail: { email }, sessionId, business } = JSON.parse(event.data.profile);
      getSettings(email, sessionId, !!business);
    }
  }

  render() {
    const {
      countries,
      dictionary,
      profile,
      regions,
      newsletter,
      displayModal,
      handleModalClose,
      modalContent,
      userDataReceived,
      application: {
        showEmailPreferences,
      },
    } = this.props;

    const {
      business,
      business: {
        accessLevel = '',
        tetheredGuid = '',
      } = {},
    } = profile;

    const appReady = userDataReceived
      && (!showEmailPreferences || !!regions.length)
      && !!countries.length
      && (!showEmailPreferences || !!Object.keys(newsletter).length)
      && !!Object.keys(dictionary).length
      && !!Object.keys(profile).length;

    const showExtrasPreferences = !(accessLevel === SELF || accessLevel === STAYER);

    return (
      <div data-test="AccountSettings">
        {appReady
        && (
          <BaseStyle className="wb-typography app-container">
            <ComponentTogglerContainer
              viewName={USER_DETAILS_VIEW}
              view={UserProfile}
              edit={UserProfileFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PASSWORD_VIEW}
              view={Password}
              edit={PasswordFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PAYMENT_CARD_VIEW}
              view={PaymentCardContainer}
              edit={PaymentCardFormContainer}
            />
            {managePermissionFlag && (
              <ManagePermission dictionary={dictionary} />
            )}
            <RegularGuestsContainer />
            {showEmailPreferences && (
              <NewsletterContainer />
            )}
            <ComponentTogglerContainer
              viewName={ROOM_REQUIREMENTS_VIEW}
              view={RoomRequirements}
              edit={RoomRequirementsFormContainer}
            />
            {showExtrasPreferences && (
              <ComponentTogglerContainer
                viewName={EXTRAS_PREFERENCES_VIEW}
                view={ExtrasPreferences}
                edit={ExtrasPreferencesFormContainer}
              />
            )}
            {!!tetheredGuid && (
              <MemorableWord />
            )}
            {business && !tetheredGuid && (
              <MoveAccount />
            )}
            {!business && (
            <DeleteCustomerContainer />
            )}
            {displayModal && <Modal data-test="Modal" display content={modalContent} handleClose={handleModalClose} />}
          </BaseStyle>
        )
      }
      </div>
    );
  }
}

App.propTypes = {
  countries: array.isRequired,
  dictionary: object.isRequired,
  profile: object.isRequired,
  getSettings: func.isRequired,
  displayModal: bool.isRequired,
  handleModalClose: func.isRequired,
  modalContent: node,
  newsletter: object.isRequired,
  regions: array.isRequired,
  userDataReceived: bool.isRequired,
  application: object.isRequired,
};

export default App;
