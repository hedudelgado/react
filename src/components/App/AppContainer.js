import { connect } from 'react-redux';
import { getSettings as getSettingsAction } from './appActions';
import { closeModal } from './modalActions';

import App from './App';

const mapStateToProps = ({
  modal: {
    display: displayModal,
    content: modalContent,
  },
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userProfile: {
    userDataReceived,
    profile,
  },
}) => ({
  profile,
  displayModal,
  modalContent,
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userDataReceived,
});

const mapDispatchToProps = dispatch => ({
  getSettings:
    (email, sessionId, business) => dispatch(getSettingsAction(email, sessionId, business)),
  handleModalClose: () => dispatch(closeModal()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
