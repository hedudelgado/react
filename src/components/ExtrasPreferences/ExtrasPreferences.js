import React from 'react';
import { object, func } from 'prop-types';
import BaseStyle from '../../layout/BaseStyle';
import Typography from '../../layout/Typography';
import Button from '../../layout/Button';
import Notification from '../../layout/Notification';
import { DEFAULT_VIEW, EXTRAS_PREFERENCES_VIEW } from '../App/appActions';

const ExtrasPreferences = ({
  app: {
    notificationMessage,
  },
  dictionary,
  setEditView,
  setNotificationMessage,
}) => {
  const showEditMode = () => {
    setEditView(EXTRAS_PREFERENCES_VIEW);
    setNotificationMessage(DEFAULT_VIEW);
  };

  return (
    <BaseStyle component="div" className="component-container">
      {notificationMessage === EXTRAS_PREFERENCES_VIEW && (
        <Notification data-test="success-message" priority={2} variant="success" ariaLive="polite" show>
          {dictionary['extraspreferences.notification.success']}
        </Notification>
      )}
      <Typography component="h4" className="pi-color-greyscale-darkest">
        {dictionary['extraspreferences.title']}
      </Typography>
      <Typography component="p" mb="3">
        {dictionary['extraspreferences.description']}
      </Typography>
      <Button data-test="edit-extras-preferences" color="quaternary" onClick={showEditMode}>
        {dictionary['extraspreferences.button.edit']}
      </Button>
    </BaseStyle>
  );
};

ExtrasPreferences.propTypes = {
  app: object.isRequired,
  dictionary: object.isRequired,
  setEditView: func.isRequired,
  setNotificationMessage: func.isRequired,
};

export default ExtrasPreferences;
