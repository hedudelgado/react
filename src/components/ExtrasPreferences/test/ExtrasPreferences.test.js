import React from 'react';
import { shallow } from 'enzyme';
import { DEFAULT_VIEW, EXTRAS_PREFERENCES_VIEW } from '../../App/appActions';
import ExtrasPreferences from '../ExtrasPreferences';

describe('ExtrasPreferences Component', () => {
  let wrapper;

  const props = {
    app: {},
    dictionary: {
      'extraspreferences.title': 'Extras Preferences',
      'extraspreferences.button.edit': 'Edit extras preferences',
    },
    setEditView: jest.fn(),
    setNotificationMessage: jest.fn(),
  };

  beforeEach(() => {
    wrapper = shallow(<ExtrasPreferences {...props} />);
  });

  describe('Renders ExtrasPreferences Component Elements', () => {
    it('should render the component', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should trigger setEditView when clicking the edit extras preferences button', () => {
      wrapper.find('[data-test="edit-extras-preferences"]').simulate('click');
      expect(props.setEditView).toHaveBeenCalledTimes(1);
      expect(props.setNotificationMessage).toHaveBeenCalledTimes(1);
    });

    it('should show success message when room requirements updated successfully', () => {
      const copyProps = {
        ...props,
        app: { editViewName: DEFAULT_VIEW, notificationMessage: EXTRAS_PREFERENCES_VIEW },
      };
      wrapper = shallow(<ExtrasPreferences {...copyProps} />);
      expect(wrapper).toContainExactlyOneMatchingElement('[data-test="success-message"]');
    });
  });
});
