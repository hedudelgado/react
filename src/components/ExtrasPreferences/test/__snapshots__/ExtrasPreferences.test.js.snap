// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtrasPreferences Component Renders ExtrasPreferences Component Elements should render the component 1`] = `
<BaseStyle
  className="component-container"
  component="div"
>
  <Typography
    className="pi-color-greyscale-darkest"
    component="h4"
  >
    Extras Preferences
  </Typography>
  <Typography
    component="p"
    mb="3"
  />
  <Button
    color="quaternary"
    data-test="edit-extras-preferences"
    fullWidth={false}
    href={null}
    onClick={[Function]}
    submit={false}
    variant="button"
  >
    Edit extras preferences
  </Button>
</BaseStyle>
`;
